rclcpp/include/rclcpp/qos_event.hpp

```diff
@@ -131,21 +132,37 @@ class QOSEventHandler : public QOSEventHandlerBase
     }
   }
 
-  /// Execute any entities of the Waitable that are ready.
+  /// Take data so that the callback cannot be called again
   void
-  execute() override
+  take_data(std::shared_ptr<void> & data) override
   {
-    EventCallbackInfoT callback_info;
+    if (data) {
+      throw std::runtime_error("Should not be data in the pointer");
```

Comment:
```suggestion
      throw std::runtime_error("'data' is not empty");
```

(also, `#include <stdexcept>` at the top for `std::runtime_error`)

Reply:
Done in https://github.com/ros2/rclcpp/pull/1241/commits/5b39f357bbe0096af397b0b782542452153ac517.